File: WdtcCore/src/main/java/org/wdt/wdtc/core/download/quilt/QuiltInstallTask.java
package org.wdt.wdtc.core.download.quilt;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import org.apache.log4j.Logger;
import org.wdt.utils.dependency.DependencyDownload;
import org.wdt.utils.gson.JsonUtils;
import org.wdt.wdtc.core.download.infterface.DownloadSourceInterface;
import org.wdt.wdtc.core.download.infterface.InstallTaskInterface;
import org.wdt.wdtc.core.download.infterface.VersionJsonObjectInterface;
import org.wdt.wdtc.core.game.GameVersionJsonObject;
import org.wdt.wdtc.core.game.Launcher;
import org.wdt.wdtc.core.game.LibraryObject;
import org.wdt.wdtc.core.manger.DownloadSourceManger;
import org.wdt.wdtc.core.utils.DownloadUtils;
import org.wdt.wdtc.core.utils.WdtcLogger;

import java.io.IOException;
import java.util.List;

public class QuiltInstallTask extends QuiltDownloadInfo implements InstallTaskInterface {

  private static final Logger logmaker = WdtcLogger.getLogger(QuiltInstallTask.class);
  private final DownloadSourceInterface source;

  public QuiltInstallTask(Launcher launcher, String quiltVersionNumber) {
    super(launcher, quiltVersionNumber);
    this.source = DownloadSourceManger.getDownloadSource();
  }

  public QuiltInstallTask(Launcher launcher, VersionJsonObjectInterface versionJsonObjectInterface) {
    super(launcher, versionJsonObjectInterface);
    this.source = DownloadSourceManger.getDownloadSource();
  }

  public void DownloadQuiltGameVersionJson() {
    DownloadUtils.StartDownloadTask(getQuiltVersionJsonUrl(), getQuiltVersionJson());
  }


  @Override
  public void overwriteVersionJson() throws IOException {
    GameVersionJsonObject VersionJsonObject = launcher.getGameVersionJsonObject();
    JsonObject QuiltVersionJsonObject = getQuiltGameVersionJsonObject();
    JsonObject Arguments = QuiltVersionJsonObject.getAsJsonObject("arguments");
    GameVersionJsonObject.Arguments arguments = VersionJsonObject.getArguments();
    JsonArray GameList = arguments.getGameList();
    VersionJsonObject.setMainClass(QuiltVersionJsonObject.get("mainClass").getAsString());
    GameList.addAll(Arguments.getAsJsonArray("game"));
    arguments.setGameList(GameList);
    VersionJsonObject.setArguments(arguments);
    JsonArray QuiltLibraryList = QuiltVersionJsonObject.getAsJsonArray("library");
    List<LibraryObject> LibraryList = VersionJsonObject.getLibraries();
    for (int i = 0; i < QuiltLibraryList.size(); i++) {
      JsonObject QuiltLibraryObject = QuiltLibraryList.get(i).getAsJsonObject();
      DependencyDownload download = new DependencyDownload(QuiltLibraryObject.get("name").getAsString());
      String LibraryDefaultUrl = QuiltLibraryObject.get("url").getAsString();
      if (LibraryDefaultUrl.equals("https://maven.fabricmc.net/")) {
        download.setDefaultUrl(source.getFabricLibraryUrl());
      } else if (LibraryDefaultUrl.equals("https://maven.quiltmc.org/repository/release/")) {
        download.setDefaultUrl("https://maven.quiltmc.org/repository/release/");
      }
      LibraryList.add(LibraryObject.getLibraryObject(download, LibraryDefaultUrl));
    }
    VersionJsonObject.setLibraries(LibraryList);
    VersionJsonObject.setId(launcher.getVersionNumber() + "-quilt-" + QuiltVersionNumber);
    launcher.putToVersionJson(VersionJsonObject);
  }

  @Override
  public void writeVersionJsonPatches() throws IOException {
    GameVersionJsonObject Object = launcher.getGameVersionJsonObject();
    Object.setJsonObject(List.of(
        JsonUtils.readFileToJsonObject(launcher.getVersionJson()),
        JsonUtils.readFileToJsonObject(getQuiltVersionJson())
    ));
    launcher.putToVersionJson(Object);
  }

  @Override
  public void afterDownloadTask() {

  }

  @Override
  public void beforInstallTask() {
    DownloadQuiltGameVersionJson();
  }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/download/quilt/QuiltInstallTask.kt
package org.wdt.wdtc.core.download.quilt

import org.wdt.utils.gson.*
import org.wdt.wdtc.core.download.infterface.ModInstallTaskInterface
import org.wdt.wdtc.core.download.infterface.VersionsJsonObjectInterface
import org.wdt.wdtc.core.game.*
import org.wdt.wdtc.core.utils.STRING_SPACE
import org.wdt.wdtc.core.utils.cleanStrInString
import org.wdt.wdtc.core.utils.startDownloadTask
import java.io.IOException

class QuiltInstallTask : QuiltDownloadInfo, ModInstallTaskInterface {
	
	constructor(version: Version, quiltVersionNumber: String) : super(version, quiltVersionNumber)
	
	constructor(version: Version, versionsJsonObjectInterface: VersionsJsonObjectInterface) : super(
		version,
		versionsJsonObjectInterface
	)
	
	override suspend fun beforInstallTask() {
		startDownloadTask(quiltVersionJsonUrl, quiltVersionJson)
	}
	
	override fun overwriteVersionJson() {
		quiltProfileJsonObject.also { quilt ->
			version.run {
				gameVersionJsonObject.apply {
					gameVersionJsonObject.apply {
						id = "${versionNumber}-quilt-$modVersion"
						mainClass = quilt.mainClass
						arguments = arguments.apply {
							jvmList = jvmList.also { game ->
								quilt.arguments.jvmList?.onEach {
									game!!.add(it.asString.cleanStrInString(STRING_SPACE))
								}
							}
							gameList = gameList.also { game ->
								quilt.arguments.gameList?.onEach {
									game!!.add(it.asString)
								}
							}
						}
						libraries.addAll(quilt.libraries.map {
							it.toLibraryObject()
						})
					}.also {
						it.putToVersionJson()
					}
					
				}
			}
		}
	}
	
	@Throws(IOException::class)
	override fun writeVersionJsonPatches() {
		version.run {
			gameVersionJsonObject.apply {
				patches = mutableListOf(
					versionJson.readFileToJsonObject(),
					quiltVersionJson.readFileToJsonObject()
				)
			}.putToVersionJson()
		}
	}
	
	override suspend fun afterDownloadTask() {}
}
